R/bs-dependencies.R

```diff
@@ -307,24 +313,37 @@ bs_dependency <- function(input = list(), theme, name, version,
 #' }
 #'
 #' @rdname bs_dependency
-bs_dependency_defer <- function(func) {
-  force(func)
+bs_dependency_defer <- function(func, memoise = TRUE) {
+  # func() most likely calls stuff like sass_file() and bs_dependency() ->
+  # sass_partial() -> sass() (e.g., see example section above) Even though
+  # sass() calls can be cached, there is still considerable overhead involved
+  # in computing the cache key itself because it involves file-level operations
+  # (e.g. getting the mtimes of file attachments).
+  # So, to help with performance (in the case where we repeatedly call the same
+  # func with the same theme), memoise func with a short-lived cache
+  if (memoise) {
+    mfunc <- memoise::memoise(func, cache = .dependency_cache)
```

Comment:
This will create a new memoized function each time `bs_dependency_defer()` is called -- so, for example, every time `selectizeInput()` is called, it will get a separate memoized function. They _do_ share the caching object, so each copy will benefit from caching, but it's extra work to do the memoization every time this function is called.

I think there may be a way to just memoize the function once, but exactly how isn't clear to me just yet...

Reply:
> it's extra work to do the memoization every time this function is called

FWIW, the overhead is 10ms in the case of example provided in #369:

<img width="1300" alt="Screen Shot 2021-11-17 at 10 49 09 AM" src="https://user-images.githubusercontent.com/1365941/142244881-5ed99c81-4d46-4fc4-81b9-7ae590137d2b.png">


It seems like getting this right is gonna make this already complicated thing much more complicated, so it doesn't seem worth it to me

Reply:
```suggestion
    # The memoized function is created each time bs_dependency_defer is called,
    # and then it is used once. This is not how memoized functions are normally
    # used, but in this case it works because the caching object is re-used, and
    # it still provides very significant improvement in performance.
    mfunc <- memoise::memoise(func, cache = .dependency_cache)
```